sql: ensure that DELETE on the fast path is still a valid data source

knz opened the pull request:
Prior to this patch, the DELETE statement without WHERE and RETURNING
would use the "fast path", that is, all the work being done in Start().

The code for this would however violate the `planNode` interface
contract: after the fast path is taken, the code would panic if the
`Next()` method is called. `planNode` specifies that `Next()` is
always callable after `Start()`, although perhaps it has nothing to
do.

Fixes #19805.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/19822)
<!-- Reviewable:end -->

knz committed:
sql: ensure that DELETE on the fast path is still a valid data source

Prior to this patch, the DELETE statement without WHERE and RETURNING
would use the "fast path", that is, all the work being done in Start().

The code for this would however violate the `planNode` interface
contract: after the fast path is taken, the code would panic if the
`Next()` method is called. `planNode` specifies that `Next()` is
always callable after `Start()`, although perhaps it has nothing to
do.

jordanlewis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> assuming you've performed analysis on all nodes with a `planNodeFastPath`!

---

Reviewed 3 of 3 files at r1.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/19822#-:-KyGnpCaWz0ZCIo6yfFG:bee0gio)*
<!-- Sent from Reviewable.io -->

knz commented:
The only non-DDL one right now is DELETE.
